diff --git a/GPU/include/edge_layer_connector.cuh b/GPU/include/edge_layer_connector.cuh
index d4854de..fec6256 100644
--- a/GPU/include/edge_layer_connector.cuh
+++ b/GPU/include/edge_layer_connector.cuh
@@ -53,9 +53,6 @@ template <class T> class CUDA_edge_layer_connector
     // GPU data
     int _blocks;
 
-    CUDA_module_map_doublet<T> _cuda_module_map_doublet;
-    CUDA_module_map_triplet<T> _cuda_module_map_triplet;
-
     void read_event (const std::string&);
 
   public:
